tensorflow_probability/python/stats/sample_stats.py

```diff
@@ -469,6 +476,81 @@ def covariance(x,
 
     return cov
 
+CovarianceState = collections.namedtuple('CovarianceState',
+                                         'num_samples, mean, comoment, dtype')
+def init_running_covariance(shape, dtype=tf.float32):
+  """Initialize the Running Covariance State.
+
+  Args:
+    shape: Python `Tuple` representing the shape of incoming samples.
+    dtype: Dtype of incoming samples and the resulting statistics.
+      By default, the dtype is `tf.float32`
+
+  Returns:
+    state: `CovarianceState` representing a stream of no inputs
+  """
+  return CovarianceState(
+      num_samples=0,
+      mean=tf.zeros(shape),
+      comoment=tf.zeros(shape + shape),
+      dtype=dtype,
+    )
+
+def update_running_covariance(curr_state, new_sample):
+  """Update the streaming statistics of the `VarianceState` with a new sample
+  using a Covariance Update Formula proposed by Philippe Pebay (2008).
+
+  Args:
+    curr_state: `CovarianceState` that depicts the current state of streaming
+      statistics
+    new_sample: Incoming sample with compatible shape and dtype as previously
+      defined in `init_running_covariance`
+
+  Returns:
+    state: `CovarianceState` with updated calculations
+
+  References:
+  [1]: Philippe Pebay. Formulas for Robust, One-Pass Parallel Computation of
+  Covariances and Arbitrary-Order Statistical Moments. _Technical Report
+  SAND2008-6212_, 2008.
+  https://prod-ng.sandia.gov/techlib-noauth/access-control.cgi/2008/086212.pdf
+  """
+  new_sample = tf.convert_to_tensor(new_sample, dtype=curr_state.dtype)
+  new_n = curr_state.num_samples + 1
+  delta_mean = new_sample - curr_state.mean
+  new_mean = curr_state.mean + delta_mean / new_n
+  all_pairwise_deltas = tf.tensordot(delta_mean, delta_mean, axes=0)
```

Comment:
Let's see whether we can get a batched version of this working.

The einsum equation for batch scalar covariance (= variance) by this algorithm is, I think,
tf.einsum("...i,...i->...i", delta_mean, delta_mean)
(I wonder whether tf.einsum("...,...->...", delta_mean, delta_mean) will also just do that.)

For batch vector covariance, I think we could use
tf.einsum("...i,...j->...ij", delta_mean, delta_mean)

For batch matrix covariance, I think we should be able to use
tf.einsum("...ij,...kl->...ijkl", delta_mean, delta_mean)

For batch over rank 3, we should be able to use
tf.einsum("...ijk,...lmn->...ijklmn", delta_mean, delta_mean)

etc.

I think everything else is transparently batch-friendly already.

WDYT?

Reply:
Should definitely be possible; in fact, the streaming variance pairwise works now (it had a subtle bug). Working on this now in addition to quantiles.